File: server/lib/javaRunner.mjs
import { execFile } from 'node:child_process'
import { mkdtemp, rm, writeFile } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import { buildJavaProject } from './harness.mjs'
import { getProblem } from './problems.mjs'

const COMPILE_TIMEOUT_MS = 20_000

function execWithTimeout(command, args, cwd, timeoutMs) {
  return new Promise((resolve, reject) => {
    const controller = timeoutMs == null ? null : new AbortController()
    const started = Date.now()
    let settled = false
    const timer =
      controller == null
        ? null
        : setTimeout(() => {
            controller.abort()
          }, timeoutMs)
    execFile(
      command,
      args,
      { cwd, signal: controller?.signal, maxBuffer: 4 * 1024 * 1024 },
      (error, stdout, stderr) => {
        if (settled) return
        settled = true
        if (timer) clearTimeout(timer)
        resolve({
          stdout,
          stderr,
          timedOut: error?.name === 'AbortError' || error?.code === 'ABORT_ERR',
          runtimeMs: Date.now() - started,
          error
        })
      }
    )
  })
}

function parseResults(stdout) {
  const results = stdout
    .split('\n')
    .filter((line) => /^(PASS|FAIL|ERROR) CASE /.test(line))
    .map((line) => {
      const match = /^(PASS|FAIL|ERROR) CASE (\d+)(.*)$/.exec(line)
      return {
        status: match[1],
        caseId: Number(match[2]),
        detail: match[3].trim()
      }
    })
  const resultMatch = /RESULT (\d+)\/(\d+)/.exec(stdout)
  return {
    results,
    pass: resultMatch ? Number(resultMatch[1]) : 0,
    total: resultMatch ? Number(resultMatch[2]) : results.length
  }
}

export async function runJava({ slug, code, timeoutMs }) {
  const problem = await getProblem(slug)
  const project = buildJavaProject(problem, code)
  const dir = await mkdtemp(join(tmpdir(), 'leetcode-run-'))
  try {
    for (const [name, source] of Object.entries(project.files)) {
      await writeFile(join(dir, name), source, 'utf8')
    }

    const compile = await execWithTimeout(
      'javac',
      Object.keys(project.files),
      dir,
      COMPILE_TIMEOUT_MS
    )
    if (compile.error) {
      return {
        pass: 0,
        total: problem.tests?.length ?? 0,
        results: [],
        stdout: compile.stdout,
        stderr: compile.stderr,
        compileError: (compile.stderr || compile.stdout).trim(),
        runtimeMs: compile.runtimeMs,
        timedOut: compile.timedOut
      }
    }

    const run = await execWithTimeout('java', ['Main'], dir, timeoutMs)
    if (run.timedOut) {
      return {
        pass: 0,
        total: problem.tests?.length ?? 0,
        results: [],
        stdout: run.stdout,
        stderr: run.stderr,
        compileError: null,
        runtimeMs: run.runtimeMs,
        timedOut: true
      }
    }

    return {
      ...parseResults(run.stdout),
      stdout: run.stdout,
      stderr: run.stderr,
      compileError: null,
      runtimeMs: run.runtimeMs,
      timedOut: false
    }
  } catch (error) {
    return {
      pass: 0,
      total: problem.tests?.length ?? 0,
      results: [],
      stdout: '',
      stderr: '',
      compileError: error.message,
      runtimeMs: 0,
      timedOut: false
    }
  } finally {
    await rm(dir, { recursive: true, force: true })
  }
}

File: src-tauri/src/java_runner.rs
use crate::harness;
use crate::problems;
use anyhow::Result;
use rand::Rng;
use std::fs;
use std::path::PathBuf;
use std::time::Instant;
use tokio::process::Command;

const COMPILE_TIMEOUT_SECS: u64 = 20;

#[derive(Debug, serde::Serialize)]
pub struct CaseResult {
    pub status: String,
    #[serde(rename = "caseId")]
    pub case_id: i64,
    pub detail: String,
}

fn parse_results(stdout: &str) -> (Vec<CaseResult>, i64, i64) {
    let mut results = Vec::new();
    let mut pass = 0i64;
    let mut total = 0i64;
    for line in stdout.lines() {
        if let Some(cap) = regex::Regex::new(r"^(PASS|FAIL|ERROR) CASE (\d+)(.*)$").unwrap().captures(line) {
            results.push(CaseResult {
                status: cap[1].to_string(),
                case_id: cap[2].parse().unwrap_or(0),
                detail: cap[3].trim().to_string(),
            });
        }
        if let Some(cap) = regex::Regex::new(r"RESULT (\d+)/(\d+)").unwrap().captures(line) {
            pass = cap[1].parse().unwrap_or(0);
            total = cap[2].parse().unwrap_or(0);
        }
    }
    if total == 0 { total = results.len() as i64; }
    (results, pass, total)
}

fn temp_dir() -> PathBuf {
    let mut p = std::env::temp_dir();
    let s: String = rand::thread_rng()
        .sample_iter(&rand::distributions::Alphanumeric)
        .take(10)
        .map(char::from)
        .collect();
    p.push(format!("leetcode-run-{}", s));
    p
}

async fn run_with_timeout(cmd: &mut Command, timeout_secs: u64) -> Result<(String, String, bool, u64)> {
    let start = Instant::now();
    let output = tokio::time::timeout(
        std::time::Duration::from_secs(timeout_secs),
        cmd.output(),
    ).await;
    let runtime_ms = start.elapsed().as_millis() as u64;
    match output {
        Ok(Ok(out)) => {
            let stdout = String::from_utf8_lossy(&out.stdout).to_string();
            let stderr = String::from_utf8_lossy(&out.stderr).to_string();
            let _timed_out = out.status.code() == Some(1); // our Main exits 1 on failure
            Ok((stdout, stderr, false, runtime_ms))
        }
        Ok(Err(e)) => Err(e.into()),
        Err(_) => Ok((String::new(), String::new(), true, runtime_ms)),
    }
}

pub async fn run_java(slug: &str, code: &str, timeout_ms: Option<u64>) -> Result<serde_json::Value> {
    let problem = problems::get_problem(slug)?;
    let timeout_ms = timeout_ms.unwrap_or(5000);
    let total = problem["tests"].as_array().map(|v| v.len()).unwrap_or(0) as i64;

    let project = harness::build_java_project(&problem, code);
    let dir = temp_dir();
    fs::create_dir_all(&dir)?;

    for (name, source) in &project.files {
        fs::write(dir.join(name), source)?;
    }

    // Compile
    let file_names: Vec<String> = project.files.keys().cloned().collect();
    let mut compile_cmd = Command::new("javac");
    compile_cmd.current_dir(&dir);
    for f in &file_names { compile_cmd.arg(f); }

    let (compile_out, compile_err, compile_timed_out, compile_ms) =
        run_with_timeout(&mut compile_cmd, COMPILE_TIMEOUT_SECS).await?;

    let compile_out_c = compile_out.clone();
    let compile_err_c = compile_err.clone();

    if compile_timed_out {
        let _ = fs::remove_dir_all(&dir);
        return Ok(serde_json::json!({
            "pass": 0, "total": total, "results": [],
            "stdout": compile_out_c, "stderr": compile_err_c,
            "compileError": "编译超时", "runtimeMs": compile_ms, "timedOut": true
        }));
    }

    // Check if compilation succeeded by looking for class files
    let compile_success = dir.join("Main.class").exists() && dir.join("Types.class").exists();
    if !compile_success {
        let err_text = if !compile_err.is_empty() { compile_err } else { compile_out };
        let _ = fs::remove_dir_all(&dir);
        return Ok(serde_json::json!({
            "pass": 0, "total": total, "results": [],
            "stdout": compile_out_c, "stderr": compile_err_c,
            "compileError": err_text.trim().to_string(),
            "runtimeMs": compile_ms, "timedOut": false
        }));
    }

    // Run
    let timeout_secs = (timeout_ms as f64 / 1000.0).max(1.0) as u64;
    let mut run_cmd = Command::new("java");
    run_cmd.current_dir(&dir).arg("Main");

    let (run_out, run_err, run_timed_out, run_ms) =
        run_with_timeout(&mut run_cmd, timeout_secs).await?;

    let _ = fs::remove_dir_all(&dir);

    if run_timed_out {
        return Ok(serde_json::json!({
            "pass": 0, "total": total, "results": [],
            "stdout": run_out, "stderr": run_err,
            "compileError": null, "runtimeMs": run_ms, "timedOut": true
        }));
    }

    let (results, pass, parsed_total) = parse_results(&run_out);
    let total = if parsed_total > 0 { parsed_total } else { total };

    Ok(serde_json::json!({
        "pass": pass, "total": total, "results": results,
        "stdout": run_out, "stderr": run_err,
        "compileError": null, "runtimeMs": run_ms, "timedOut": false
    }))
}
